core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1237,6 +1237,57 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * Finializers of each inner stream are garuenteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When either the inner or outer stream fails, the entire streams fails.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    for {
+      haltRef <- Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None))
+      haltOuter <- Stream.eval(Deferred[F2, Unit])
+      queue <- Stream.eval(fs2.async.unboundedQueue[F2, Option[Either[Throwable, Chunk[O2]]]])
+      pulse <- Stream.eval(fs2.async.unboundedQueue[F2, Option[(O, Deferred[F2, Unit])]])
+      guard <- Stream.eval(Semaphore(1))
+      o2 <- queue.dequeue.unNoneTerminate.rethrow
+        .flatMap(Stream.chunk(_))
+        .concurrently(
+          pulse.dequeue.unNoneTerminate
+            .onFinalize(queue.enqueue1(None))
+            .evalMap {
+              case (o, haltInner) =>
+                guard.withPermit(
+                  f(o).chunks
+                    .evalMap(x => queue.enqueue1(Right(x).some))
+                    .interruptWhen(haltInner.get.attempt)
+                    .interruptWhen(haltOuter.get.attempt)
+                    .compile
+                    .drain
+                    .handleErrorWith(e => haltOuter.complete(()) *> queue.enqueue1(Left(e).some)))
+            }
+            .merge(this.attempt
+              .interruptWhen(haltOuter.get.attempt)
+              .onFinalize(pulse.enqueue1(None))
+              .evalMap {
+                case Left(e) =>
+                  haltRef.modify(x => None -> x.fold(F2.unit)(_.complete(()))).flatten *>
+                    guard.withPermit(queue.enqueue1(Left(e).some))
+                case Right(x) =>
+                  Deferred[F2, Unit].flatMap { halt =>
+                    haltRef.modify {
+                      case None       => halt.some -> F2.unit
+                      case Some(last) => halt.some -> last.complete(())
+                    }.flatten *> pulse.enqueue1((x -> halt).some)
+                  }
+              }))
+    } yield o2
```

Comment:
Minor optimization: the final `yield o2` will result in a `map(o2 => o2)` call, which will then result in chunk copying (each chunk will have `.map(o2 => o2)` called on it). To fix, we could manually expand the for-comprehension.

Reply:
probably true, but I tend not to use `for` with streams anyway, because you end needing a lot of methods on stream (e.g. `evalMap`, `concurrently`, `to`, `take`) and so on, and with explicit `flatMap` is easier to mix them in that with `for`